File: src/invoke/workspace.ts
import { invoke } from '@tauri-apps/api/core';

export type WorkspaceState = {
	rootPath: string | null;
	openTabPaths: string[];
	lastActiveFilePath: string | null;
	sidebarWidth: number | null;
	sortEnabled: boolean | null;
	showHiddenFiles: boolean | null;
	tabBarMode: string | null;
	zoomLevel: number | null;
};

/** Get the persisted workspace state (called on app start). */
export async function getWorkspaceState(): Promise<WorkspaceState> {
	return invoke<WorkspaceState>('get_workspace_state');
}

/** Set workspace root path (clears tabs when different from current). */
export async function setWorkspaceRoot(rootPath: string | null): Promise<void> {
	return invoke('set_workspace_root', { rootPath });
}

/** Add a file to the open tabs list. */
export async function addTab(filePath: string): Promise<void> {
	return invoke('add_tab', { filePath });
}

/** Remove a file from the open tabs list. */
export async function closeTab(filePath: string): Promise<void> {
	return invoke('close_tab', { filePath });
}

/** Remove multiple files from the open tabs list. */
export async function closeTabs(filePaths: string[]): Promise<void> {
	return invoke('close_tabs', { filePaths });
}

/** Set the active (focused) tab. */
export async function setActiveTab(filePath: string | null): Promise<void> {
	return invoke('set_active_tab', { filePath });
}

/** Persist sidebar width. */
export async function setSidebarWidth(width: number): Promise<void> {
	return invoke('set_sidebar_width', { width });
}

/** Persist tab bar mode ("scroll" | "wrap"). */
export async function setTabBarMode(mode: string): Promise<void> {
	return invoke('set_tab_bar_mode', { mode });
}

/** Persist webview zoom level (1.0 = 100%). */
export async function setZoomLevel(zoomLevel: number): Promise<void> {
	return invoke('set_zoom_level', { zoomLevel });
}

/** Replace the open tab paths wholesale. */
export async function setOpenTabPaths(paths: string[]): Promise<void> {
	return invoke('set_open_tab_paths', { paths });
}

/** Clear all persisted workspace state. */
export async function clearWorkspaceState(): Promise<void> {
	return invoke('clear_workspace_state');
}

/**
 * Resolve a markdown image source to an absolute filesystem path.
 * The caller should then pass the result through `convertFileSrc()`
 * to obtain a Tauri asset protocol URL.
 */
export async function resolveImageSrc(
	src: string,
	filePath: string,
	rootPath: string | null
): Promise<string> {
	return invoke<string>('resolve_image_src', { src, filePath, rootPath });
}

File: src-tauri/src/commands/workspace.rs
use crate::models::workspace::WorkspaceState;
use crate::protocol::MadoraProtocolState;
use crate::services::workspace::WorkspaceStore;
use tauri::State;

#[tauri::command]
pub async fn get_workspace_state(
    store: State<'_, WorkspaceStore>,
) -> Result<WorkspaceState, String> {
    store.get_state()
}

#[tauri::command]
pub async fn set_workspace_root(
    store: State<'_, WorkspaceStore>,
    protocol_state: State<'_, MadoraProtocolState>,
    root_path: Option<String>,
) -> Result<(), String> {
    // Sync the workspace root into the protocol handler state
    protocol_state.set_workspace_root(root_path.as_ref().map(std::path::PathBuf::from));
    store.set_root_path(root_path)
}

#[tauri::command]
pub async fn add_tab(store: State<'_, WorkspaceStore>, file_path: String) -> Result<(), String> {
    store.add_tab(&file_path)
}

#[tauri::command]
pub async fn close_tab(store: State<'_, WorkspaceStore>, file_path: String) -> Result<(), String> {
    store.close_tab(&file_path)
}

#[tauri::command]
pub async fn close_tabs(
    store: State<'_, WorkspaceStore>,
    file_paths: Vec<String>,
) -> Result<(), String> {
    store.close_tabs(&file_paths)
}

#[tauri::command]
pub async fn set_active_tab(
    store: State<'_, WorkspaceStore>,
    file_path: Option<String>,
) -> Result<(), String> {
    store.set_active_tab(file_path.as_deref())
}

#[tauri::command]
pub async fn set_sidebar_width(store: State<'_, WorkspaceStore>, width: u32) -> Result<(), String> {
    store.set_sidebar_width(width)
}

#[tauri::command]
pub async fn set_tab_bar_mode(
    store: State<'_, WorkspaceStore>,
    mode: String,
) -> Result<(), String> {
    store.set_tab_bar_mode(&mode)
}

#[tauri::command]
pub async fn set_zoom_level(
    store: State<'_, WorkspaceStore>,
    zoom_level: f64,
) -> Result<(), String> {
    store.set_zoom_level(zoom_level)
}

#[tauri::command]
pub async fn set_open_tab_paths(
    store: State<'_, WorkspaceStore>,
    paths: Vec<String>,
) -> Result<(), String> {
    store.set_open_tab_paths(&paths)
}

#[tauri::command]
pub async fn clear_workspace_state(store: State<'_, WorkspaceStore>) -> Result<(), String> {
    store.clear()
}

/// Resolve a markdown image source to an absolute filesystem path.
///
/// The frontend should call `convertFileSrc()` on the returned path
/// to obtain a Tauri asset protocol URL.
#[tauri::command]
pub async fn resolve_image_src(
    src: String,
    file_path: String,
    root_path: Option<String>,
) -> Result<String, String> {
    Ok(crate::services::workspace::resolve_image_src(
        &src,
        &file_path,
        root_path.as_deref(),
    ))
}
